Realm/RLMSyncSubscription.mm

```diff
@@ -0,0 +1,483 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2021 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#import "RLMSyncSubscription_Private.hpp"
+#import "RLMRealm_Private.hpp"
+#import "RLMUtil.hpp"
+#import "RLMQueryUtil.hpp"
+#import "RLMObjectId_Private.hpp"
+#import "RLMSyncUtil.h"
+
+#import <realm/sync/subscriptions.hpp>
+#import <realm/status_with.hpp>
+#import <realm/util/future.hpp>
+
+#pragma mark - Subscription
+
+@interface RLMSyncSubscription () {
+    std::unique_ptr<realm::sync::Subscription> _subscription;
+    RLMSyncSubscriptionSet *_subscriptionSet;
+}
+@end
+
+@implementation RLMSyncSubscription
+
+- (instancetype)initWithSubscription:(realm::sync::Subscription)subscription subscriptionSet:(RLMSyncSubscriptionSet *)subscriptionSet {
+    if (self = [super init]) {
+        _subscription = std::make_unique<realm::sync::Subscription>(subscription);
+        _subscriptionSet = subscriptionSet;
+        return self;
+    }
+    return nil;
+}
+
+- (RLMObjectId *)identifier {
+    return [[RLMObjectId alloc]initWithValue:_subscription->id()];
+}
+
+- (nullable NSString *)name {
+    const std::string_view str_view = _subscription->name();
+    std::string str = std::string(str_view);
+    const char * characters = str.c_str();
+    NSString *name = [NSString stringWithCString:characters
+                                        encoding:[NSString defaultCStringEncoding]];
+    if ([name length] == 0) {
+        return NULL;
+    }
+    return name;
```

Comment:
```suggestion
    auto strView = _subscription->name();
    if (strView.size() == 0) {
        return nil;
    }
    return [[NSString alloc] initWithBytes:strView.data()
                                    length:strView.size()
                                  encoding:NSUTF8StringEncoding];
```

Converting the std::string_view to a std::string doesn't do anything useful. `[NSString defaultCStringEncoding]` does complicated things which are incorrect here; strings passed to and from core must always be UTF-8. Passing the size is required to support embedded nuls in the name; if we don't want to support that then it can be just `return @(strView.data());`